rfcs/012-add-session-ssh-support/012-add-session-ssh-support.md

```diff
@@ -0,0 +1,438 @@
+
+Other solutions could be envisaged: it could be possible, for example, to
+provide a tunnel over HTTP/S into the session and access it via ssh locally or
+perhaps some wireguard solution might work but these approaches would probably
+encounter issues with different OS versions, would not be easy to configure and
+ultimately would likely have a detrimental impact on user experience and hence
+are not considered further.
+
+### Exposing ssh port from each session directly
+
+In this approach, each user session pod exposes at least 2 ports: an HTTP port
+for the Jupyter lab server and an ssh port for the ssh session. These ports
+have corresponding services running inside the pod.
+
+Kubernetes services are then linked to these sessions with the ports on the
+services mapping to the ports exposed by the pod. An ingress is then mapped to
+the ssh services with a dedicated ingress for each ssh session.
+
+Each ingress should have a unique name and also a unique IP address as the ssh
+protocol operates on an IP address basis (i.e., it has no specific server
+identification as noted above). This solution can suffer from DNS propagation
+delays, depending on the DNS provider and DNS configuration. Further, exposing
+ssh servers running inside user sessions directly to the Internet brings some
+security risks and it would need to be clear that users cannot easily change
+the ssh daemon configuration within their sessions to make their session very
+exposed.
+
+A couple of further points relating to this approach:
+- ingresses on cloud providers can typically be expensive; having one per
+  session is likely to incur significant cost;
+- on the Switch deployments, the entire cluster is exposed via a single IP
+  address; as such it is not straightforward to devise a solution in which
+  different sessions would map to different IP addresses.
+
+### Adding a proxy jumphost
+
+Another solution and that which is currently deployed is to have a jumphost and
+to use this to access the pods. In this case, a jumphost must be installed
+inside the kubernetes cluster as it requires direct access to the ssh server
+running on the pods. The jumphost has a standard ssh server running with proxying
+enabled.
+
+For the Switch/Openstack deployments, this requires modifications to the
+cluster ingress configuration to forward TCP traffic incident on a specific
+port to the jumphost pod on its ssh port. For cloud providers, a new ingress
+pointing at the ssh proxy is straightforward. ssh proxying approaches require
+authentication against both the proxy and the destination server; this
+necessitates an approach in which some combination of no credentials and a
+keypair must be deployed to the proxy and/or user session.
+
+The most natural solution in this case is to use the user's public key within
+the session and have open access with no ability to obtain a shell within the
+jumphost. Key management then becomes a user responsibility, with users having
```

Comment:
Not necessarily - we could inject ssh keys into the session for the user, once there is a possibility for us to store the public keys.

Reply:
Sure - will amend accordingly.